File: packages/scan/src/noise-reduction/despeckle.algorithm.ts
/**
 * Median-filter the R, G and B channels of an RGBA buffer; alpha is copied.
 *
 * A median, unlike the mean the background is estimated with, rejects the
 * isolated light and dark specks a scanner or photocopier leaves instead of
 * smearing them into their neighbours - which, once contrast is stretched, is
 * how a speck fuses into a thin stroke and a `1` becomes a `l`.
 *
 * Each output is the median of the square window of `radius` around it,
 * clipped at the image border - the element at `length >> 1` of the sorted
 * window, so a clipped window of even size takes the upper middle. For the
 * common 3x3 window, interior pixels use a 19-comparison selection network
 * (Paeth) instead of a sort: the same value, several times faster, on a page
 * of millions of pixels times three channels.
 */
export function despeckle (data: Uint8ClampedArray, width: number, height: number, radius = 1): Uint8ClampedArray {
  const r = Math.max(1, Math.round(radius))
  const out = new Uint8ClampedArray(data.length)
  const side = 2 * r + 1
  const window = new Uint8Array(side * side)
  const nine = new Uint8Array(9)

  for (let y = 0; y < height; y++) {
    const y0 = Math.max(0, y - r)
    const y1 = Math.min(height - 1, y + r)

    for (let x = 0; x < width; x++) {
      const index = (y * width + x) * 4
      out[index + 3] = data[index + 3]
      const interior = r === 1 && y > 0 && y < height - 1 && x > 0 && x < width - 1

      for (let c = 0; c < 3; c++) {
        if (interior) {
          const above = index - width * 4 + c
          const here = index + c
          const below = index + width * 4 + c
          nine[0] = data[above - 4]
          nine[1] = data[above]
          nine[2] = data[above + 4]
          nine[3] = data[here - 4]
          nine[4] = data[here]
          nine[5] = data[here + 4]
          nine[6] = data[below - 4]
          nine[7] = data[below]
          nine[8] = data[below + 4]
          out[index + c] = median9(nine)
          continue
        }

        const x0 = Math.max(0, x - r)
        const x1 = Math.min(width - 1, x + r)
        let n = 0
        for (let ny = y0; ny <= y1; ny++)
          for (let nx = x0; nx <= x1; nx++) window[n++] = data[(ny * width + nx) * 4 + c]
        out[index + c] = selectMiddle(window, n)
      }
    }
  }

  return out
}

/** The sorted element at `n >> 1` of the first `n` values, by insertion sort - windows here are tiny. */
function selectMiddle (values: Uint8Array, n: number): number {
  for (let i = 1; i < n; i++) {
    const v = values[i]
    let j = i - 1
    while (j >= 0 && values[j] > v) {
      values[j + 1] = values[j]
      j--
    }
    values[j + 1] = v
  }

  return values[n >> 1]
}

/** Median of nine values by Paeth's selection network; leaves `p` partly sorted. */
function median9 (p: Uint8Array): number {
  sort(p, 1, 2)
  sort(p, 4, 5)
  sort(p, 7, 8)
  sort(p, 0, 1)
  sort(p, 3, 4)
  sort(p, 6, 7)
  sort(p, 1, 2)
  sort(p, 4, 5)
  sort(p, 7, 8)
  sort(p, 0, 3)
  sort(p, 5, 8)
  sort(p, 4, 7)
  sort(p, 3, 6)
  sort(p, 1, 4)
  sort(p, 2, 5)
  sort(p, 4, 7)
  sort(p, 4, 2)
  sort(p, 6, 4)
  sort(p, 4, 2)

  return p[4]
}

function sort (p: Uint8Array, a: number, b: number): void {
  if (p[a] <= p[b]) return

  const t = p[a]
  p[a] = p[b]
  p[b] = t
}

File: python-packages/scanmate-scan/scanmate_scan/noise_reduction/despeckle_algorithm.py
"""Median-filter the R, G and B channels of an RGBA image; alpha is copied.

A median, unlike the mean the background is estimated with, **rejects** the
isolated light and dark specks a scanner or photocopier leaves instead of
smearing them into their neighbours - which, once contrast is stretched, is how
a speck fuses into a thin stroke and a ``1`` becomes an ``l``.

Each output is the median of the square window of ``radius`` around it, clipped
at the image border - the element at ``length >> 1`` of the sorted window, so a
clipped window of even size takes the **upper** middle.
"""

from __future__ import annotations

import math

import numpy
from numpy.typing import NDArray


def despeckle(
    pixels: NDArray[numpy.uint8], width: int, height: int, radius: float = 1
) -> NDArray[numpy.uint8]:
    """Median-filter an RGBA image.

    :param pixels: The image, shaped ``(height, width, 4)``.
    :param width: Its width, for the same signature the TypeScript has.
    :param height: Its height.
    :param radius: Half the window's side. A number, as the TypeScript takes:
        rounded with Math.round semantics and held at 1 or more.
    :returns: A new image of the same shape.
    """
    r = max(1, _js_round(radius))
    out = numpy.empty((height, width, 4), dtype=numpy.uint8)
    # Alpha is copied, never filtered: a median of the alpha channel would
    # invent edges where a page is transparent.
    out[:, :, 3] = pixels[:, :, 3]

    for channel in range(3):
        out[:, :, channel] = _median_channel(pixels[:, :, channel], width, height, r)

    return out


def _median_channel(
    channel: NDArray[numpy.uint8], width: int, height: int, r: int
) -> NDArray[numpy.uint8]:
    """One channel, median-filtered with the window clipped at the border."""
    side = 2 * r + 1
    # The interior, where every window is full, is done with one stacked
    # median: the same value the TypeScript's selection network gives, because
    # a full window of odd size has one unambiguous middle.
    out = numpy.empty((height, width), dtype=numpy.uint8)
    if height > 2 * r and width > 2 * r:
        stack = numpy.empty((side * side, height - 2 * r, width - 2 * r), dtype=numpy.uint8)
        at = 0
        for dy in range(side):
            for dx in range(side):
                stack[at] = channel[dy : dy + height - 2 * r, dx : dx + width - 2 * r]
                at += 1
        stack.sort(axis=0)
        out[r : height - r, r : width - r] = stack[(side * side) >> 1]

    # The border, where the window is clipped and may hold an even count - and
    # where the middle is therefore the UPPER one, which a mean-based median
    # would get wrong by half a level.
    for y in range(height):
        inside_rows = r <= y < height - r
        y0 = max(0, y - r)
        y1 = min(height - 1, y + r)
        for x in range(width):
            if inside_rows and r <= x < width - r:
                continue
            x0 = max(0, x - r)
            x1 = min(width - 1, x + r)
            window = channel[y0 : y1 + 1, x0 : x1 + 1].reshape(-1)
            ordered = numpy.sort(window, kind="stable")
            out[y, x] = ordered[len(ordered) >> 1]

    return out


def _js_round(value: float) -> int:
    """``Math.round``: halves go up, not to even."""
    return math.floor(value + 0.5)
